chicago/permit_cleaning.py

```diff
@@ -0,0 +1,237 @@
+in separate csvs in separate folders.
+"""
+
+import requests
+import pandas as pd
+from sodapy import Socrata
+from datetime import datetime, timedelta
+import os
+
+
+# DEFINE FUNCTIONS
+
+def download_all_permits():
+    # update limit in url below when ready to work with full dataset (as of Nov 17, 2023 dataset has 756,766 rows)
+    url = "https://data.cityofchicago.org/resource/ydr8-5enu.json?$limit=5000"
+    permits_response = requests.get(url)
+    print("status code: ", permits_response.status_code)
+    print("Headers: ", permits_response.headers.get("Content-Type"))
+    permits = permits_response.json()
+    permits_df = pd.DataFrame(permits)
+    return permits_df
+
+
+def expand_multi_pin_permits(df):
+    """ 
+    Data from the Chicago open data permits table (data this script works with) has rows uniquely identified by permit number.
+    Permits can apply to multiple PINs, with additional PINs recorded in the PIN2 - PIN10 fields.
+    We want rows that are uniquely identified by PIN and permit number. 
+    This function creates new rows for each additional PIN in multi-PIN permits and saved the relevant PIN in pin_solo.
+    """    
+    # the downloaded dataframe will not include any pin columns that are completely blank, so check for existing ones here
+    all_pin_columns = ["pin1", "pin2", "pin3", "pin4", "pin5", "pin6", "pin7", "pin8", "pin9", "pin10"]
+    pin_columns = [col for col in df.columns if col in all_pin_columns]
+    extra_pins = [pin for pin in pin_columns if pin != "pin1"]
+    non_pin_columns = [col for col in df.columns if col not in pin_columns]
+
+    print("number of not NA values in pin2 - pin7 fields: ", df[extra_pins].notna().sum().sum())
+    print("starting number of rows: ", len(df)) 
+    melted_df = pd.melt(df, id_vars=non_pin_columns, value_vars=pin_columns, var_name="pin_type", value_name="solo_pin")
+    print("number of rows after melting before removing NaNs: ", len(melted_df))
+    
+    # keep rows with NA for pin1, filter out rows with NA for other pins
+    melted_df = melted_df[(melted_df["pin_type"] == "pin1") | ((melted_df["pin_type"] != "pin1") & melted_df["solo_pin"].notna())]
+    print("number of rows after removing NaNs: ", len(melted_df))
+    
+    # order rows by permit number then pin type (so pins will be in order of their assigned numbering in permit table, not necessarily by pin number)
+    melted_df = melted_df.sort_values(by=["permit_", "pin_type"]).reset_index(drop=True)
+    
+    return melted_df
+
+
+# update pin to match formatting of iasWorld
+def format_pin(df): 
+    # iasWorld format doesn't include dashes
+    df["pin_final"] = df["solo_pin"].astype(str).str.replace("-", "")
+    # add zeros to 10-digit PINs to transform into 14-digits PINs
+    df["pin_final"] = df["pin_final"].apply(lambda x: x + "0000" if len(x) == 10 else x if x != "nan" else "")
```

Comment:
I'm using x != "nan" here instead of notna() since I converted the PIN variable to a string, so NaNs became"nan" strings. notna() still works better throughout the rest of the script, though, I think.